File: windows/Models/AppSettings.cs
using System.IO;
using System.Text.Json;

namespace QuadClicker.Models;

/// <summary>User preferences persisted to %APPDATA%\QuadClicker\settings.json.</summary>
public sealed class AppSettings
{
    // ── Click rate ────────────────────────────────────────────────────────────
    // Mode determines which set of units is valid for ClickRateUnit.
    //   Delay:     "ms", "sec", "min"
    //   Frequency: "per_sec", "per_min", "per_hour"
    public ClickRateMode ClickRateMode  { get; set; } = ClickRateMode.Delay;
    public string        ClickRateValue { get; set; } = "100";
    public string        ClickRateUnit  { get; set; } = "ms";

    // ── Click behavior ────────────────────────────────────────────────────────
    public MouseButton Button             { get; set; } = MouseButton.Left;
    public ClickType   ClickType          { get; set; } = ClickType.Single;
    public bool        UseCurrentPosition { get; set; } = true;
    public int         X                  { get; set; } = 0;
    public int         Y                  { get; set; } = 0;
    public int         StopAfterClicks    { get; set; } = 0;
    public double      StopAfterSeconds   { get; set; } = 0;
    public double      IdleWaitSeconds    { get; set; } = 0;
    public bool        AlwaysOnTop        { get; set; } = false;
    public string      StartHotkeyText    { get; set; } = "";
    public string      StopHotkeyText     { get; set; } = "F10";

    // ── Persistence ───────────────────────────────────────────────────────────

    private static readonly string SettingsPath = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
        "QuadClicker", "settings.json");

    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };

    public static AppSettings Load()
    {
        try
        {
            if (File.Exists(SettingsPath))
                return LoadFromJson(File.ReadAllText(SettingsPath));
        }
        catch { /* Corrupt or unreadable — fall back to defaults */ }
        return new AppSettings();
    }

    /// <summary>Deserializes JSON and runs forward-only legacy migration. Public for tests.</summary>
    public static AppSettings LoadFromJson(string json)
    {
        var settings = JsonSerializer.Deserialize<AppSettings>(json, JsonOptions) ?? new AppSettings();
        settings.MigrateLegacy();
        return settings;
    }

    public void Save()
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(SettingsPath)!);
            File.WriteAllText(SettingsPath, JsonSerializer.Serialize(this, JsonOptions));
        }
        catch { /* Non-fatal — settings loss is recoverable */ }
    }

    /// <summary>
    /// Translates pre-redesign unit values ("/s", "/min") into the new (Mode, Unit) shape.
    /// Idempotent: canonical values pass through unchanged.
    /// </summary>
    private void MigrateLegacy()
    {
        switch (ClickRateUnit)
        {
            case "/s":
                ClickRateMode = ClickRateMode.Frequency;
                ClickRateUnit = "per_sec";
                break;
            case "/min":
                ClickRateMode = ClickRateMode.Frequency;
                ClickRateUnit = "per_min";
                break;
            case "ms":
                ClickRateMode = ClickRateMode.Delay;
                break;
        }
    }
}

File: macos/QuadClicker/Models/AppSettings.swift
// Models/AppSettings.swift
// QuadClicker — macOS
//
// Persisted to ~/Library/Application Support/QuadClicker/settings.json.
// JSON keys match the Windows implementation for cross-platform compatibility.

import Foundation

final class AppSettings: Codable {
    // ── Click Rate ────────────────────────────────────────────────────────────
    // Mode determines which set of unit tags is valid for clickRateUnit.
    //   .delay     → "ms", "sec", "min"
    //   .frequency → "per_sec", "per_min", "per_hour"
    var clickRateMode: ClickRateMode = .delay
    var clickRateValue: String       = "100"
    var clickRateUnit: String        = "ms"

    // ── Click Behaviour ───────────────────────────────────────────────────────
    var button: MouseButton   = .left
    var clickType: ClickType  = .single

    // ── Location ──────────────────────────────────────────────────────────────
    var useCurrentPosition: Bool = true
    var x: Int = 0
    var y: Int = 0

    // ── Stop Conditions ───────────────────────────────────────────────────────
    var stopAfterClicks: Int    = 0
    var stopAfterSeconds: Double = 0
    var idleWaitSeconds: Double  = 0

    // ── Window ────────────────────────────────────────────────────────────────
    var alwaysOnTop: Bool = false

    // ── Hotkeys ───────────────────────────────────────────────────────────────
    var startHotkeyText: String = ""
    var stopHotkeyText: String  = "F10"

    // ── Coding keys (match Windows JSON property names) ───────────────────────
    enum CodingKeys: String, CodingKey {
        case clickRateMode       = "ClickRateMode"
        case clickRateValue      = "ClickRateValue"
        case clickRateUnit       = "ClickRateUnit"
        case button              = "Button"
        case clickType           = "ClickType"
        case useCurrentPosition  = "UseCurrentPosition"
        case x                   = "X"
        case y                   = "Y"
        case stopAfterClicks     = "StopAfterClicks"
        case stopAfterSeconds    = "StopAfterSeconds"
        case idleWaitSeconds     = "IdleWaitSeconds"
        case alwaysOnTop         = "AlwaysOnTop"
        case startHotkeyText     = "StartHotkeyText"
        case stopHotkeyText      = "StopHotkeyText"
    }

    init() {}

    // Custom init from decoder so missing `ClickRateMode` doesn't fail decoding.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        clickRateMode      = (try? c.decode(ClickRateMode.self, forKey: .clickRateMode))     ?? .delay
        clickRateValue     = (try? c.decode(String.self,        forKey: .clickRateValue))    ?? "100"
        clickRateUnit      = (try? c.decode(String.self,        forKey: .clickRateUnit))     ?? "ms"
        button             = (try? c.decode(MouseButton.self,   forKey: .button))            ?? .left
        clickType          = (try? c.decode(ClickType.self,     forKey: .clickType))         ?? .single
        useCurrentPosition = (try? c.decode(Bool.self,          forKey: .useCurrentPosition))?? true
        x                  = (try? c.decode(Int.self,           forKey: .x))                 ?? 0
        y                  = (try? c.decode(Int.self,           forKey: .y))                 ?? 0
        stopAfterClicks    = (try? c.decode(Int.self,           forKey: .stopAfterClicks))   ?? 0
        stopAfterSeconds   = (try? c.decode(Double.self,        forKey: .stopAfterSeconds))  ?? 0
        idleWaitSeconds    = (try? c.decode(Double.self,        forKey: .idleWaitSeconds))   ?? 0
        alwaysOnTop        = (try? c.decode(Bool.self,          forKey: .alwaysOnTop))       ?? false
        startHotkeyText    = (try? c.decode(String.self,        forKey: .startHotkeyText))   ?? ""
        stopHotkeyText     = (try? c.decode(String.self,        forKey: .stopHotkeyText))    ?? "F10"
        migrateLegacy()
    }

    // ── Migration: legacy "/s" / "/min" → canonical Frequency tags ────────────
    private func migrateLegacy() {
        switch clickRateUnit {
        case "/s":
            clickRateMode = .frequency
            clickRateUnit = "per_sec"
        case "/min":
            clickRateMode = .frequency
            clickRateUnit = "per_min"
        case "ms":
            clickRateMode = .delay
        default:
            break
        }
    }

    // ── Persistence ───────────────────────────────────────────────────────────

    private static var settingsURL: URL {
        let appSupport = FileManager.default.urls(
            for: .applicationSupportDirectory, in: .userDomainMask
        ).first!
        return appSupport
            .appendingPathComponent("QuadClicker", isDirectory: true)
            .appendingPathComponent("settings.json")
    }

    static func load() -> AppSettings {
        let url = settingsURL
        guard FileManager.default.fileExists(atPath: url.path) else {
            return AppSettings()
        }
        do {
            let data = try Data(contentsOf: url)
            let decoder = JSONDecoder()
            return try decoder.decode(AppSettings.self, from: data)
        } catch {
            return AppSettings()
        }
    }

    func save() {
        let url = Self.settingsURL
        do {
            let dir = url.deletingLastPathComponent()
            try FileManager.default.createDirectory(
                at: dir, withIntermediateDirectories: true, attributes: nil
            )
            let encoder = JSONEncoder()
            encoder.outputFormatting = .prettyPrinted
            let data = try encoder.encode(self)
            try data.write(to: url, options: .atomic)
        } catch {
            // Non-fatal — settings loss is recoverable
        }
    }
}
